server/src/main/java/org/elasticsearch/action/admin/indices/recovery/TransportRecoveryAction.java

```diff
@@ -107,4 +111,19 @@ protected ClusterBlockException checkGlobalBlock(ClusterState state, RecoveryReq
     protected ClusterBlockException checkRequestBlock(ClusterState state, RecoveryRequest request, String[] concreteIndices) {
         return state.blocks().indicesBlockedException(ClusterBlockLevel.METADATA_READ, concreteIndices);
     }
+
+    @Nullable // unless running tests that inject extra behaviour
+    private volatile Runnable onShardOperation;
```

Comment:
This is totally stupid, but I just can't see a better way to block these operations for the purposes of this test.